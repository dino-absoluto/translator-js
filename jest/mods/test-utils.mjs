/**
 * @author Dino <dinoabsoluto+dev@gmail.com>
 * @license
 * This file is part of translator-js.
 *
 * translator-js is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * translator-js is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with translator-js.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
/* eslint-env jest */
/* imports */
import path from 'path'
import globby from 'globby'
import hasha from 'hasha'
import makeDir from 'make-dir'
/* -imports */

export const globshot = async (...argv) => {
  let files = await globby(...argv)
  files = files.sort()
  files = files.map(async fname => {
    let content = await hasha.fromFile(fname, {
      encoding: 'base64'
    })
    return {
      fname,
      content
    }
  })
  files = await Promise.all(files)
  return files
}

export const setupChdir = (fname) => {
  const __rootDir = process.cwd()
  const __tmpdir = path.resolve(fname)
  beforeEach(() => {
    makeDir.sync(__tmpdir)
    process.chdir(__tmpdir)
  })
  afterEach(() => {
    process.chdir(__rootDir)
  })
}
