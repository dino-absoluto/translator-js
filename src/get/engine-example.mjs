/**
 * @file Get module
 * @author Dino <dinoabsoluto+dev@gmail.com>
 * @license
 * This file is part of translator-js.
 *
 * translator-js is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * translator-js is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with translator-js.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
/* imports */
import * as base from './base'
/* -imports */

export class Chapter extends base.Chapter {
  update () {
    const { props } = this
    const files = [
      new base.FileInfo({
        chapter: this,
        fname: this.getName(`${props.title}.txt`),
        integrity: undefined,
        buffer: `${props.title}\n---\n\nHello World!`
      })
    ]
    this.setProps({ files })
  }
}

export class Volume extends base.Volume {
}

export class Series extends base.Series {
  get Chapter () {
    return Chapter
  }

  get Volume () {
    return Volume
  }

  refresh () {
    const volumes = [
      {
        title: 'The Beginning'
      },
      {
        title: 'The Happening'
      }
    ]
    const chapters = [
      {
        title: 'Prologue',
        integrity: 'prologue',
        volume: 0
      },
      {
        title: 'Reborn',
        integrity: 'reborn',
        volume: 1
      }
    ]
    this.updateData({
      volumes,
      chapters
    })
  }
}
