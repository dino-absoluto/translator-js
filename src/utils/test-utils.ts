/**
 * @author Dino <dinoabsoluto+dev@gmail.com>
 * @license
 * Copyright 2019 Dino <dinoabsoluto+dev@gmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
/* imports */
import globby, { GlobbyOptions } from 'globby'
import * as path from 'path'
import del from 'del'
import { back as nockBack, NockBackContext } from 'nock'
import hasha = require('hasha')
import makeDir = require('make-dir')

/* setup */
const SRC_DIR = path.normalize(__dirname + '/../')
const TMP_DIR = path.resolve('__tmp__/jest/')
const FIXTURES_DIR = path.resolve('__tmp__/nock-fixtures')
nockBack.fixtures = FIXTURES_DIR
nockBack.setMode('record')

/* code */
export const setupTmpDir = (name: string) => {
  const fpath = path.join(TMP_DIR, name)
  beforeAll(async () => {
    await del(path.join(fpath, '*'))
    await makeDir(fpath)
  })
  return fpath
}

interface TestEnvSetupOptions {
  network?: boolean
}

export const setup = (fname: string, options: TestEnvSetupOptions = {}) => {
  const unit = path.posix.normalize(path.relative(SRC_DIR, fname))
    .replace(/\/([^\/]+)\/__tests__\//g, '/$1__')
  const tmpdir = path.join(TMP_DIR, unit)
  let nock: { nockDone: () => void; context: NockBackContext }
  beforeAll(async () => {
    await del(path.join(tmpdir, '*'))
    await makeDir(tmpdir)
    if (options.network) {
      nock = await nockBack(unit + '.json')
    }
  })
  afterAll(async () => {
    if (nock && nock.nockDone) {
      await nock.nockDone()
    }
  })
  return {
    __TMPDIR: tmpdir
  }
}

export const hashDir = async (
  patt: string | ReadonlyArray<string>,
  options: GlobbyOptions
) => {
  let files = await globby(patt, options)
  files = files.sort()
  const cwd = options.cwd || path.resolve('.')
  const arrays = await Promise.all(files.map(async (fname: string) => {
    let content = await hasha.fromFile(fname, {
      encoding: 'base64'
    })
    return {
      fname: path.posix.normalize(path.relative(cwd, fname)),
      content
    }
  }))
  const data: { [id: string]: string } = {}
  for (const { fname, content } of arrays) {
    data[fname] = content || ''
  }
  return data
}

export const hash = async (data: any) => {
  let text: string
  if (typeof data === 'string') {
    text = data
  } else {
    text = JSON.stringify(data)
  }
  return hasha(text, {
    encoding: 'base64'
  })
}
