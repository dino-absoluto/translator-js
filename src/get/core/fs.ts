/**
 * @author Dino <dinoabsoluto+dev@gmail.com>
 * @license
 * Copyright 2019 Dino <dinoabsoluto+dev@gmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
/* imports */
import { trim, flow } from '../../utils/flow'
import * as pfs from '../../utils/pfs'
import * as path from 'path'
import makeDir = require('make-dir')
/* code */

const sanitizeName = (name?: string): string | undefined => {
  return flow(trim(name))
    .then(path.normalize)
    .then((name): string | undefined => {
      if (path.basename(name) !== name) {
        return
      }
      if (name.startsWith('..')) {
        return
      }
      return name
    }).value
}

interface FSItem {
  parent?: Folder
  path: string
  real (): Promise<void>
  remove (): Promise<void>
  close (): Promise<void>
  rename (name: string): Promise<void>
}

export class File implements FSItem {
  public readonly parent: Folder
  private _filename?: string

  public constructor (parent: Folder, name: string) {
    this.parent = parent
    this._filename = sanitizeName(name)
  }

  public get path (): string {
    const { parent, _filename } = this
    if (!_filename) {
      throw new Error('filename is not set')
    }
    return path.join(parent.path, _filename)
  }

  public async real (): Promise<void> {
    if (!this._filename) {
      throw new Error('filename is not set')
    }
  }

  public async rename (name: string): Promise<void> {
    const newName = sanitizeName(name)
    if (!newName) {
      throw new Error('filename is invalid')
    }
    const oldName = this._filename
    this._filename = newName
    if (oldName) {
      if (newName === oldName) {
        return
      }
      try {
        const parentDir = this.parent.path
        await pfs.rename(
          path.join(parentDir, oldName),
          path.join(parentDir, newName))
      } catch (err) {
        if (err.code !== 'ENOENT') {
          throw err
        }
      }
    }
  }

  public async close (): Promise<void> {
    delete this._filename
    this.parent.children.delete(this)
  }

  public async access (mode?: number): Promise<void> {
    return pfs.access(this.path, mode)
  }

  public async remove (): Promise<void> {
    await this.real()
    const fpath = this.path
    await this.close()
    try {
      await pfs.unlink(fpath)
    } catch (err) {
      if (err.code !== 'ENOENT') {
        throw err
      }
    }
  }

  public async read (): Promise<Buffer> {
    await this.real()
    return pfs.readFile(this.path)
  }

  public async write (content: string | Buffer): Promise<void> {
    await this.parent.real()
    await this.real()
    await pfs.writeFile(this.path, content)
  }
}

export class Folder implements FSItem {
  public readonly children: Set<Folder | File> = new Set()
  public readonly parent?: Folder
  /* private data */
  private _dirname?: string
  private _accessed = false

  public constructor (parent: Folder | undefined | null, name: string) {
    if (parent) {
      this.parent = parent
      this._dirname = sanitizeName(name)
    } else {
      this._dirname = path.resolve(name)
    }
  }

  public get name (): string | undefined { return this._dirname }
  public get path (): string {
    const { parent, _dirname } = this
    if (!_dirname) {
      throw new Error('Folder\'s path is undefined')
    }
    if (parent) {
      return path.join(parent.path, _dirname)
    }
    return _dirname
  }

  public async real (): Promise<void> {
    if (this._accessed) {
    } else {
      await makeDir(this.path)
      this._accessed = true
    }
  }

  public async close (): Promise<void> {
    delete this._dirname
    if (this.parent) {
      this.parent.children.delete(this)
    }
  }

  public async remove (): Promise<void> {
    const { children } = this
    if (!children.size) {
      throw new Error('folder is not empty')
    }
    const fpath = this.path
    await this.close()
    await pfs.rmdir(fpath)
  }

  public get renameable (): boolean { return !!this.parent }
  public async rename (name: string): Promise<void> {
    if (!this.parent) {
      throw new Error('folder cannot be renamed')
    }
    const newName = sanitizeName(name)
    if (!newName) {
      throw new Error('folder name is invalid')
    }
    const oldName = this._dirname
    this._dirname = newName
    this._accessed = false
    if (oldName) {
      if (newName === oldName) {
        return
      }
      try {
        const parentDir = this.parent.path
        await pfs.rename(
          path.join(parentDir, oldName),
          path.join(parentDir, newName))
      } catch (err) {
        if (err.code !== 'ENOENT') {
          throw err
        }
      }
    }
  }

  public requestFolder (name: string): Folder {
    const entry = new Folder(this, name)
    this.children.add(entry)
    return entry
  }

  public requestFile (name: string): File {
    const entry = new File(this, name)
    this.children.add(entry)
    return entry
  }
}
