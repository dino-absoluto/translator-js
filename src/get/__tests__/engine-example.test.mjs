/**
 * @file Get module
 * @author Dino <dinoabsoluto+dev@gmail.com>
 * @license
 * This file is part of translator-js.
 *
 * translator-js is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * translator-js is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with translator-js.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
/* eslint-env jest */
/* imports */
import * as Engine from '../engine-example'
import fs from 'fs'
import del from 'del'
import path from 'path'
import makeDir from 'make-dir'
import * as utils from 'test-utils'
/* -imports */

utils.setupChdir('__tmp__/tests/example')

const initData = async (prefix) => {
  await del(prefix)
  await makeDir(`${prefix}/00 Chapter One`)
  const sourceURL = new URL('https://www.example.com')
  const write = (fname, data) => {
    fs.writeFileSync(`${prefix}/${fname}`, data)
  }
  write('index.json', JSON.stringify({
    sourceURL,
    volumes: [
      {
        'title': 'Chapter One'
      }
    ],
    chapters: [
      {
        title: 'Description',
        integrity: 'start',
        volume: 0,
        files: [{
          fname: '000 Description.txt',
          integrity: undefined
        }]
      }
    ]
  }, null, 1))
  write('00 Chapter One/000 Description.txt', 'Hello World!')
}

test('simple', async () => {
  const prefix = 'simple'
  await initData(prefix)
  let series = new Engine.Series({
    source: prefix
  })
  series = await series.refresh()
  expect(series.targetDir).toBe(path.resolve(prefix))
  expect(await utils.globshot(prefix)).toMatchSnapshot()
})
