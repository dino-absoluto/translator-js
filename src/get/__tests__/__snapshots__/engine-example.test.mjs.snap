// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`simple 1`] = `
Array [
  Object {
    "content": "hhhE1nBOhXP+w02WfiC8/vPUJM9IvgTm3AjyvVjHKXQzcQFerYkcw88cnTS0kmS1EHUbH/nlN5N7xGtdb/TsyA==",
    "fname": "simple/00 Chapter One/000 Description.txt",
  },
  Object {
    "content": "KPOANbK0dyH6PffJq+LuHiODTuhV+Dfecq6PmzUKPNq44ajFIhfgygys/MbwwXq8/udnAHjREyXhqEcxYXeocg==",
    "fname": "simple/01 The Beginning/001 Prologue.txt",
  },
  Object {
    "content": "H5a27bIHyc3o73aFBbj5+7GxLqcCTVzVcOhDJKgfdROpuB9plKd6mxD9IsQTbREtHeeO3Og62S9zhEpI8cb54A==",
    "fname": "simple/01 The Beginning/002 Prologue 2.txt",
  },
  Object {
    "content": "kzsVsAC8QOIkYMfjV32AYif6TYHbPe1mOr3QUJHVrdeaNT3qkdVN39rwl0iFxp4Cs/S1YUhKwdcrZq5wH8XzXw==",
    "fname": "simple/index.json",
  },
]
`;
